### PR TITLE
Delta-5 Content ID > Backend > Email Notification For User Applications

### PR DESCRIPTION
**Pull request by FerrielMelarpis**
#### Changes
- Added .gitignore
- Added _.toUpper on language key passed in derive_language method.

#### Ticket References:
- [Delta-5](https://freedom.myjetbrains.com/youtrack/issue/Delta-5)

> These changes are required for anyTV/content-id-service#31

### CODE REVIEW

**Review by pandafox — COMMENTED** [review]
Add a test case? The current example is actually using lower case strings

**Comment by FerrielMelarpis** [comment]
What sample? 
The current test case uses `language()` method which is different from `derive_language()`. 
Also, I kind of need help implementing the test case.
If you look at the test for this library, it's not an actual unit test but rather use case test.

**Comment by pandafox** [comment]
https://github.com/anyTV/anytv-templater/blob/be3498f354768b28be13c0339cb1d82bfefe0c73/test/test.js#L58

And also in the intro
https://github.com/anyTV/anytv-templater#introduction

Hum, perhaps we're doing something wrong?

**Comment by FerrielMelarpis** [comment]
Again... That's `language()` method which sets the default language. You need to pass lowercase for that because that is the one being mapped to language property of country_language_map
https://github.com/anyTV/anytv-templater/blob/be3498f354768b28be13c0339cb1d82bfefe0c73/src/config/country_language_map.js#L3-L6

**Comment by pandafox** [comment]
Right, but looks like we shouldn’t be doing upper because we need to be able to lookup by two letter as well we the full country name?

@wildtron

**Comment by FerrielMelarpis** [comment]
Hmmm, The `_.toUpper` is there for the lookup callback passed in `_.filter` https://github.com/anyTV/anytv-templater/pull/8/files#diff-0bec058c9142d049091518af8e923ecfL105

Who's jet btw?

**Comment by pandafox** [comment]
jet is sherwin lol.. typing at jet automatically convert to @wildtron on web, but not on mobile

https://github.com/anyTV/anytv-templater/pull/8/files#diff-0bec058c9142d049091518af8e923ecfL97
This is the part where we lookup by both the country code + country name

**Comment by FerrielMelarpis** [comment]
Yup... and keys of coutries is UPPERCASE hence the need for `_.toUpper`. Once we have a match, we use  the `language` property of that map e.g. ('en').

**Comment by pandafox** [comment]
Yeah but if always do toUpper then if the input is the country's name then it won't ever match, right? So I guess we'll have to either only do toUpper when the string length is 2 (since there's no country's full name is 2 character)

**Comment by FerrielMelarpis** [comment]
I don't get it really... derive_language is different from language method.... 
language sets the country code whilst derive_language looks for country code from the country_language_map config

**Comment by pandafox** [comment]
Yeah I'm talking about derive_language, it takes a list of country code, or country name. So we cannot just use toUpper case (since it only apply to 2 letter country code)

File: src/Templater.js [hunk 1]
```diff
@@ -94,7 +94,7 @@ export default class Templater {
         const find = key =>
             language_group =>
                 _.includes(language_group.countries, key)
-                || language_group.countries[key];
+                || language_group.countries[_.toUpper(key)];
```
Comment (FerrielMelarpis) [inline src/Templater.js]: @pandafox How about we restrict the toUpper() only here so we don't affect other use cases?

Comment (pandafox) [reply src/Templater.js]: I think that'll work, but it'll mean we need to convert multiple times for the same value.
Prob more efficient if we just convert to 2 chars.

As for testing, we can just test the value of _language after calling derived_language with
1. lower case 2 letter
2. upper case 2 letter
3. all lower case country name
4. proper case country name

etc.,

Comment (FerrielMelarpis) [reply src/Templater.js]: hmmm I think we can do that using spies. I'll see what I can do.

File: test/Templater.spec.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+
+chai.should();
+chai.use(sinonChai);
+
+
+describe('#Templater', function () {
+
+
+    describe('#Templater.derive_language', function () {
```
Comment (FerrielMelarpis) [inline test/Templater.spec.js:19]: @pandafox Please check if we still need other test cases.

File: src/Templater.js [hunk 3]
```diff
@@ -89,13 +89,17 @@ export default class Templater {
...
+    derive_language(...keys) {
+
+        const find = key => {
 
-        const find = key =>
-            language_group =>
-                _.includes(language_group.countries, key)
-                || language_group.countries[key];
+            const country_code = _.upperCase(key);
+            const country_name = _.startCase(_.lowerCase(key));
```
Comment (pandafox) [inline src/Templater.js:97]: startCase should already taken care of the lowerCase
https://lodash.com/docs/4.17.4#startCase
https://en.wikipedia.org/wiki/Letter_case#Stylistic_or_specialised_usage

File: test/Templater.spec.js (hunk 2 above)
Comment (pandafox) [reply test/Templater.spec.js:19]: The test cases looks great.

Perhaps add a test case for empty string, empty array

**Review by pandafox — CHANGES_REQUESTED** [review]

File: src/Templater.js (hunk 3 above)
Comment (FerrielMelarpis) [reply src/Templater.js:97]: I don't think so.
https://jsfiddle.net/ferrielmelarpis/ytafe24o/
Also, if you check the implementation, it doesn't apply lowercase to trailing letters of a word.
https://github.com/lodash/lodash/blob/462bc9bf3b0ac2738098d3045e80a2a8910203de/startCase.js#L24-L28

Comment (pandafox) [reply src/Templater.js:97]: Okay, guess I read it wrong, thought all cap is the input in the wiki's example.

**Review by pandafox — APPROVED** [review]

**Review by wilbertverayin — APPROVED** [review]
